### ISSUE TITLE
add predator prey model

### ISSUE THREAD

**Issue by fdschneider** [fdschneider/caspr#9]
Create predprey.R. Transfer C code from  [here](https://github.com/skefi/SpatialStress/blob/master/OldCodes/PredPrey/PredPrey.C) into update function

**Comment by fdschneider** [comment]
Use the [model_template.R](https://github.com/fdschneider/caspr/blob/master/R/model_template.R) and plug in an update function that wraps around the C code using [Rcpp](http://dirk.eddelbuettel.com/code/rcpp.html).

**Comment by alexgenin** [comment]
I added it in pull request #22, but it needs more testing to check whether it produces good results.

**Comment by fdschneider** [comment]
Thanks Alex. I am checking it out. Amazing how much faster that is. 
We should think about transferring the other models to C++ as well! I'll have a look. It would slim down the package quite a bit, since we can omit the `count()`  and `mapping()` function.

**Comment by fdschneider** [comment]
There was a bug with the backtransformation of the matrix into a factorial vector. This mixed up the attribution of states after each timestep and caused nonsensical updates. 

Also, I set the number of evaluations (subs) to width*height, which gives each cell the chance to be updated at the globally set rate. otherwise the parameter values would have different impact at different sizes of the lattice.

Now it seems to work well. Try: 

```

l <- init_landscape(c("f","s","0"), c(0.6,0.01,0.4), 100, 100)
p <- list()
p$betaf = 0.99
p$betas = 0.05
p$delta = 0.01
r <- ca(l, predprey, parms = p, saveeach = 1, t_min = 300, t_max = 400, t_eval = 400, steady = 0.00001)
animate(r, "predprey.gif")
```

Although I have the impression that the predators don't reliably go extinct if no resource is present. Shouldn't they all starve to death? We need to verify the model outcome. 

Please don't continue work on the `compiled_code` branch. I merged everything into master!

**Comment by alexgenin** [comment]
I get this from your snippet:
![image](http://alex.lecairn.org/files/predprey.gif)

Yeah the predators should go extinct eventually if no resources are present, I was trying to replicate the results from Fig. 1 in the original publication (Pascual 2002): 

![image](http://alex.lecairn.org/files/fig1.png)

They suggest betaf = 1/3, betas = 1/10 and delta = 1/3 (the \nu parameter does not appear in the code), but I get many more predators for these parameters. Will investigate :)

**Comment by fdschneider** [comment]
Arg. we should add a warning for epileptics to this issue!

Are you sure you were running the newest caspr? the psychedelic flickering is what I got from running your original code, because it mixed up the cell states after each update when transferring from numerical into factorial. I thought I fixed that in c4c317ce15094e1349f77c78c45deab8cb436e77. 

But anyhow, even with the default parameters, my fish go extinct pretty fast.

**Comment by alexgenin** [comment]
You're right I was still on the compiled_code branch, hence the flickering. It's fixed now, thanks ! 

I'll investigate the model results on tuesday.

**Comment by alexgenin** [comment]
I'm reopening this issue as this model produces strange output. To be completely sure we should try to reproduce Figure. 1 (see above), but that requires a patch-size distribution fitting function. Will investigate.

**Comment by alexgenin** [comment]
OK, I fixed the obvious wrong behavior in 14a9000c55131ddbf9be6532b22268f9d7c59459. Now we just need to decide on a sensible set of parameters.

**Comment by fdschneider** [comment]
excellent. I modified the defaults to a set that produces some interesting pattern, closing this for now.
